src/cdomains/arrayDomain.ml

```diff
@@ -941,7 +994,746 @@ struct
+      if Z.lt max_i min_size && all_indexes_must_null min_i max_i then
+        Null
+        (* if maximum value in index interval < maximal size, return NotNull if no number in index interval is in may_nulls_set *)
+      else if Z.lt max_i max_size && not (MayNulls.exists (fun x -> Z.geq x min_i && Z.leq x max_i) may_nulls_set) then
+        NotNull
+      else
+        Top
+    (* if maximum number in interval is invalid, i.e. negative, return Top of value *)
+    | _ -> Top
+
+  (* helper functions *)
+  let must_nulls_remove i must_nulls_set min_size =
+    let rec compute_set acc i =
+      if Z.geq i min_size then 
+        acc
+      else
+        compute_set (MustNulls.add i acc) (Z.succ i) in
+    if MustNulls.is_bot must_nulls_set then
+      MustNulls.remove i (compute_set (MustNulls.empty ()) Z.zero)
+    else
+      MustNulls.remove i must_nulls_set
+  let must_nulls_filter cond must_nulls_set min_size = 
+    let rec compute_set acc i =
+      if Z.geq i min_size then 
+        acc
+      else
+        compute_set (MustNulls.add i acc) (Z.succ i) in
+    if MustNulls.is_bot must_nulls_set then
+      MustNulls.filter cond (compute_set (MustNulls.empty ()) Z.zero)
+    else
+      MustNulls.filter cond must_nulls_set
+  let must_nulls_min_elt must_nulls_set =
+    if MustNulls.is_bot must_nulls_set then
+      Z.zero
+    else
+      MustNulls.min_elt must_nulls_set
+  let may_nulls_remove i may_nulls_set max_size =
+    let rec compute_set acc i =
+      if Z.geq i max_size then
+        acc
+      else
+        compute_set (MayNulls.add i acc) (Z.succ i) in
+    if MayNulls.is_top may_nulls_set then
+      MayNulls.remove i (compute_set (MayNulls.empty ()) Z.zero)
+    else
+      MayNulls.remove i may_nulls_set
+  let may_nulls_filter cond may_nulls_set max_size =
+    let rec compute_set acc i =
+      if Z.geq i max_size then
+        acc
+      else
+        compute_set (MayNulls.add i acc) (Z.succ i) in
+    if MayNulls.is_top may_nulls_set then
+      MayNulls.filter cond (compute_set (MayNulls.empty ()) Z.zero)
+    else
+      MayNulls.filter cond may_nulls_set
+  let may_nulls_min_elt may_nulls_set =
+    if MayNulls.is_top may_nulls_set then
+      Z.zero
+    else
+      MayNulls.min_elt may_nulls_set
+
+  let set (ask: VDQ.t) (must_nulls_set, may_nulls_set, size) (e, i) v =
+    let rec add_indexes i max may_nulls_set =
+      if Z.gt i max then
+        may_nulls_set
+      else
+        add_indexes (Z.succ i) max (MayNulls.add i may_nulls_set) in
+    let min interval = match Idx.minimal interval with
+      | Some min_num when Z.geq min_num Z.zero -> min_num
+      | _ -> Z.zero in (* assume worst case minimal natural number *)
+
+    let min_size = min size in
+    let min_i = min i in
+    let max_i = idx_maximal i in
+
+    let set_exact i =
+      match idx_maximal size with
+      (* if size has no upper limit *)
+      | None -> 
```

Comment:
You explained to me at some point why `Val.is_not_null` is not needed, but I am now confused about this again: It seems there are three cases:

- Value is definitely `\0`
- Value is definitely not `\0`
- Value may be `\0` or not

I don't quite follow why you only need a boolean function here...

Reply:
Ah, I think I messed up because I implemented the domain before implementing the `is_null` function in `Compound` of *valueDomain*, assuming that `is_null` returning `false` would mean it's definitely not `\0`. For my assumption, no `is_not_null` was needed. Of course, that changes things. I'll add a `is_not_null` function after all.